File: miriego-backend/app/models/notificacion.py
"""
Modelos ORM (SQLAlchemy) para el módulo de Notificaciones.
Reflejan las tablas creadas en db/notificaciones_schema.sql.
"""

import enum
from datetime import datetime, timezone

from sqlalchemy import (
    Boolean,
    Column,
    Integer,
    Sequence,
    String,
    Text,
    DateTime,
    Enum as SAEnum,
)

from app.core.database import Base

SCHEMA = "miriego"


class NotificadoTipo(str, enum.Enum):
    regante = "regante"
    tercero = "tercero"


class EstadoNotificacion(str, enum.Enum):
    emitida = "emitida"
    notificada = "notificada"
    respondida = "respondida"
    vencida = "vencida"
    cumplida = "cumplida"
    cerrada = "cerrada"


class TipoNotificacion(Base):
    __tablename__ = "tipos_notificaciones"
    __table_args__ = {"schema": SCHEMA}

    id = Column(Integer, primary_key=True)
    nombre = Column(String(150), nullable=False)
    activo = Column(Boolean, nullable=False, default=True)


class MedioNotificacion(Base):
    __tablename__ = "medios_notificacion"
    __table_args__ = {"schema": SCHEMA}

    id = Column(Integer, primary_key=True)
    nombre = Column(String(100), nullable=False)
    activo = Column(Boolean, nullable=False, default=True)


class Notificacion(Base):
    __tablename__ = "notificaciones"
    __table_args__ = {"schema": SCHEMA}

    id = Column(Integer, primary_key=True)
    codigo_notificacion = Column(String(30), nullable=False, unique=True)

    tipo_notificacion_id = Column(Integer)
    medio_notificacion_id = Column(Integer)
    expediente_id = Column(Integer)

    notificado_tipo = Column(
        SAEnum(NotificadoTipo, name="notificado_tipo", schema=SCHEMA),
        nullable=False,
        default=NotificadoTipo.tercero,
    )
    cc = Column(String(20))
    pp = Column(String(20))

    notificado_nombre = Column(String(200))
    notificado_documento = Column(String(30))
    notificado_domicilio = Column(Text)
    notificado_contacto = Column(String(150))

    motivo = Column(String(200), nullable=False)
    descripcion = Column(Text, nullable=False)

    fecha_emision = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    fecha_notificacion = Column(DateTime(timezone=True))
    fecha_vencimiento_respuesta = Column(DateTime(timezone=True))

    estado = Column(
        SAEnum(EstadoNotificacion, name="estado_notificacion", schema=SCHEMA),
        nullable=False,
        default=EstadoNotificacion.emitida,
    )

    usuario_id = Column(Integer, nullable=False, default=1)
    observaciones = Column(Text)

    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

File: miriego-frontend/src/lib/types/notificacion.ts
// Tipos TypeScript para el módulo de Notificaciones.
// Reflejan los schemas Pydantic del backend (app/schemas/notificacion.py).

export type NotificadoTipo = 'regante' | 'tercero';

export type EstadoNotificacion =
	| 'emitida'
	| 'notificada'
	| 'respondida'
	| 'vencida'
	| 'cumplida'
	| 'cerrada';

// --- Catálogos ---

export interface TipoNotificacion {
	id: number;
	nombre: string;
}

export interface MedioNotificacion {
	id: number;
	nombre: string;
}

// --- Notificación ---

export interface Notificacion {
	id: number;
	codigo_notificacion: string;
	usuario_id: number;

	tipo_notificacion_id?: number | null;
	medio_notificacion_id?: number | null;
	expediente_id?: number | null;

	notificado_tipo: NotificadoTipo;
	cc?: string | null;
	pp?: string | null;
	inspeccion_id?: number | null;
	notificado_nombre?: string | null;
	notificado_documento?: string | null;
	notificado_domicilio?: string | null;
	notificado_contacto?: string | null;

	motivo: string;
	descripcion: string;

	fecha_emision?: string | null;
	fecha_notificacion?: string | null;
	fecha_vencimiento_respuesta?: string | null;

	estado: EstadoNotificacion;
	observaciones?: string | null;

	numero_expediente?: string | null;
}

export interface NotificacionDetalle extends Notificacion {
	inspeccion_nombre?: string | null;
	inspector_nombre?: string | null;
}

export interface NotificacionCreatePayload {
	tipo_notificacion_id?: number | null;
	medio_notificacion_id?: number | null;
	expediente_id?: number | null;
	notificado_tipo: NotificadoTipo;
	cc?: string;
	pp?: string;
	inspeccion_id?: number;
	notificado_nombre?: string;
	notificado_documento?: string;
	notificado_domicilio?: string;
	notificado_contacto?: string;
	motivo: string;
	descripcion: string;
	fecha_vencimiento_respuesta?: string;
	observaciones?: string;
}

export interface PaginatedResponse<T> {
	items: T[];
	total: number;
	page: number;
	page_size: number;
}
